### PR TITLE
gossiper: wait_for_live_nodes_to_show_up: allow grace for pending alive endpoints

### PR DESCRIPTION
**Pull request by bhalevy**
There are spurious dtest failure where
wait_for_live_nodes_to_show_up times up after 30 seconds waiting for 2 nodes to show up in gossip.

I'm seeing this locally when running multiple topology-related dtest with -n 6 parallelism and I managed to reproduce this by adding a sleep of 10 seconds before real_mark_alive in the send_gossip_echo continuation.

This commit extends the loop timeout if a change is seen in _pending_mark_alive_endpoints.  This means that there's a `mark_alive` in progress and possibly
`real_mark_alive` may be blocking in lock_endpoint.

Still, limit the loop to max_timeout of 6 minutes
(twice the debug-mode timeout).

### CODE REVIEW

**Comment by bhalevy** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3003/

Unrelated. https://github.com/scylladb/scylladb/issues/14809

File: gms/gossiper.cc [hunk 1]
```diff
@@ -2309,11 +2312,20 @@ future<> gossiper::wait_for_live_nodes_to_show_up(size_t n) {
     constexpr auto timeout_delay = std::chrono::seconds{30};
 #endif
     auto timeout = gossiper::clk::now() + timeout_delay;
+    auto max_timeout = gossiper::clk::now() + std::chrono::minutes{6};
+    std::unordered_set<inet_address> pending_mark_alive_endpoints;
     while (get_live_members().size() < n) {
-        if (timeout <= gossiper::clk::now()) {
+        auto now = gossiper::clk::now();
+        if (timeout <= now) {
+          if (_pending_mark_alive_endpoints != pending_mark_alive_endpoints && max_timeout > now) {
```
Comment (kbr-scylla) [inline gms/gossiper.cc:2320]: Please also add a commit which fixes indentation

File: gms/gossiper.cc [hunk 2]
```diff
@@ -2309,11 +2312,20 @@ future<> gossiper::wait_for_live_nodes_to_show_up(size_t n) {
...
+            logger.info("pending_mark_alive_endpoints changed while waiting for live nodes: {}", _pending_mark_alive_endpoints);
```
Comment (kbr-scylla) [inline gms/gossiper.cc:2321]: Note that if we reach timeout, this is always going to be printed, because `pending_mark_alive_endpoints` is initialized to `nullopt`. So even if `_pending_mark_alive_endpoints` doesn't change during this `timeout_delay`, the message will incorrectly say that it did.

I think we should initialize it to `_pending_mark_alive_endpoints`.

File: gms/gossiper.cc (hunk 1 above)
Comment (bhalevy) [reply gms/gossiper.cc:2320]: no problem

File: gms/gossiper.cc (hunk 2 above)
Comment (bhalevy) [reply gms/gossiper.cc:2321]: It's going to be printed if `_pending_mark_alive_endpoints` is non-null at timeout time.
But I agree it would be more elegant to initialize `pending_mark_alive_endpoints` to `_pending_mark_alive_endpoints` so to detect the first change, if any, more accurately.

**Comment by bhalevy** [comment]
In v2 ([61930bc](https://github.com/scylladb/scylladb/commit/61930bc42e995ec1fa4d95271fe8300bb1ff8ec8)):
- Initialize `pending_mark_alive_endpoints` to `_pending_mark_alive_endpoints`
- Fixup indentation

**Comment by nyh** [comment]
@bhalevy the commit message talks about a "spurious dtest failure", but the fix isn't to dtest but to the Scylla code, which suggests it isn't "spurious" but a real bug... So was this a Scylla bug, where Scylla gives up too quickly, or a test bug that the test gives up too quickly?

Your commit message has a lot of details but I find it hard to understand the bigger picture of what is being fixed here and why.

File: gms/gossiper.cc [hunk 3]
```diff
@@ -1550,7 +1550,7 @@ void gossiper::mark_alive(inet_address addr, endpoint_state& local_state) {
     // Enter the _background_msg gate so stop() would wait on it
     auto gh = _background_msg.hold();
     logger.debug("Sending a EchoMessage to {}, with generation_number={}", id, generation);
-    (void)_messaging.send_gossip_echo(id, generation.value(), std::chrono::milliseconds(15000)).then([this, addr] {
+    (void)_messaging.send_gossip_echo(id, generation.value(), std::chrono::milliseconds(15000)).then([this, addr] () -> future<> {
```
Comment (nyh) [inline gms/gossiper.cc:1553]: Some of the changes in this patch, like this one and the other one below, aren't really relevant and I'm not sure if they are justified.
Perhaps if you want them, you can move them to a separate patch with some excuse why you want them?

**Comment by bhalevy** [comment]
The dtest framework is paranoid in the sense that it fails when hitting unexpected error in the log.
Those are checked after the test business logic has run, regardless if it passed or failed.
In this case, the test business logic found no issue and it was just the dtest framework that failed the test based on the unexpected error.
We can filter out this error to make the test pass, but I'm afraid this is not the right thing to do since the view builder already ignores `sleep_aborted` as it is expected on shutdown and therefore not an error.
The same should apply to `abort_requested_exception` and this is what this patch is about.
I don't know what exactly triggered the error and what changed in scylla or seastar that we got that one instead of the expected `sleep_aborted`.

File: gms/gossiper.cc [hunk 4]
```diff
@@ -2311,6 +2311,9 @@ future<> gossiper::wait_alive(std::vector<gms::inet_address> nodes, std::chrono:
 }
 
 future<> gossiper::wait_for_live_nodes_to_show_up(size_t n) {
+    if (this_shard_id() != 0) {
```
Comment (nyh) [inline gms/gossiper.cc:2314]: Why this change (and again, it's not really part of the main change)?
This function doesn't work on other shards? Why?

File: gms/gossiper.cc [hunk 5]
```diff
@@ -2311,6 +2311,9 @@ future<> gossiper::wait_alive(std::vector<gms::inet_address> nodes, std::chrono:
...
+        on_internal_error(logger, "Must be called on shard 0");
```
Comment (nyh) [inline gms/gossiper.cc:2315]: It's  better to have unique assertion-failure messages that you can grep without looking at full backtraces. For example, this message can be something like "gossiper::wait_for_live_nodes_to_show_up must be called on shard 0".

File: gms/gossiper.cc [hunk 6]
```diff
@@ -2311,6 +2311,9 @@ future<> gossiper::wait_alive(std::vector<gms::inet_address> nodes, std::chrono:
...
+    }
     logger::rate_limit rate_limit{std::chrono::seconds{5}};
 #ifdef SEASTAR_DEBUG
     // Account for debug slowness. 3 minutes is probably overkill but we don't want flaky tests.
```
Comment (nyh) [inline gms/gossiper.cc:2319]: Nitpick: This comments talks about "3 minutes" but below you also add another 6 minute timeout. Would have been nice to have a comment which clarifies what all these numbers mean together.

**Review by nyh — COMMENTED** [review]
In general I think this patch is ok, but:
1. I left some comments on specific things I thought could be improved, 
2. I don't understand why you try so hard to separate the timeout between waiting for one next node and the timeout for all the nodes. Is it really more realistic/usual that we get individual nodes appearing 30 seconds apart than two appearing together after 1 minute? What would we lose if we don't add ANY of this complexity, and just increase the old 3-minute timeout that already existed in the code to 6 minutes?

In any case I defer to @kbr-scylla 's judgement on this PR. He is the one that added that 3-minute thing originally (fd66bb1a61e735106938c8b5797e848caa8737ad) so he has better insights on what he was trying to do there.

File: gms/gossiper.cc (hunk 3 above)
Comment (bhalevy) [reply gms/gossiper.cc:1553]: sorry, debugging leftovers, will remove hunk.

File: gms/gossiper.cc (hunk 5 above)
Comment (bhalevy) [reply gms/gossiper.cc:2315]: okay

File: gms/gossiper.cc (hunk 4 above)
Comment (bhalevy) [reply gms/gossiper.cc:2314]: `_pending_mark_alive_endpoints` is updated only on shard 0 hence the need.
That said, we can use invoke_on to run the body explicitly on shard 0, but since it is called only on shard 0 in practice, we can just assert that.

File: gms/gossiper.cc (hunk 6 above)
Comment (bhalevy) [reply gms/gossiper.cc:2319]: okay

**Comment by bhalevy** [comment]
> @bhalevy the commit message talks about a "spurious dtest failure", but the fix isn't to dtest but to the Scylla code, which suggests it isn't "spurious" but a real bug... So was this a Scylla bug, where Scylla gives up too quickly, or a test bug that the test gives up too quickly?

The problem this patch fixes is on scylla's side for giving up too quickly.
One factor contributing to the flakiness is that we're waiting for other nodes to act as for other nodes to appear as a live they need to gossip to us.

I think that the algorithm could be made more robust and use much tighter timing if we proactively "ping"ed other nodes to get their endpoint_state, maybe those we learnt about in the shadow round, and send them an unsolicited ack message rather than wait until they send us a "syn" message.

> 
> Your commit message has a lot of details but I find it hard to understand the bigger picture of what is being fixed here and why.

The bigger picture is that the bootstrapping node might passively wait for a long time for other nodes to appear as alive locally, and that times out in low probability and based on the load and timing on other nodes.

**Comment by bhalevy** [comment]
> In general I think this patch is ok, but:
> 
> 1. I left some comments on specific things I thought could be improved,
> 2. I don't understand why you try so hard to separate the timeout between waiting for one next node and the timeout for all the nodes. Is it really more realistic/usual that we get individual nodes appearing 30 seconds apart than two appearing together after 1 minute? What would we lose if we don't add ANY of this complexity, and just increase the old 3-minute timeout that already existed in the code to 6 minutes?
> 
> In any case I defer to @kbr-scylla 's judgement on this PR. He is the one that added that 3-minute thing originally ([fd66bb1](https://github.com/scylladb/scylladb/commit/fd66bb1a61e735106938c8b5797e848caa8737ad)) so he has better insights on what he was trying to do there.

I don't know, maybe simply increasing the timeout to 3 minutes or more in all modes (not just debug) would be enough for all practical purposes.

And the comprehensive fix could be what I described in https://github.com/scylladb/scylladb/pull/15016#issuecomment-1676314156 which is proactively "ping"ing other nodes to get their endpoint_state and trigger the message exchange that eventually marks them as alive locally, rather than waiting passively for gossip to run its coarse.

**Comment by bhalevy** [comment]
In v3 ([fc55ebf](https://github.com/scylladb/scylladb/commit/fc55ebf5f1cd459566bf86a6ddca65d96450154d)):
- Improve on_internal_error message, comments, and commit log message

**Comment by bhalevy** [comment]
https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/6408/artifact/dtest/logs-full.release.005/1692196572172_counter_tests.py%3A%3ATestCountersOnMultipleNodes%3A%3Atest_counter_consistency_node_repair/node3.log
```
INFO  2023-08-16 14:35:13,193 [shard 0] gossip - Waiting for 2 live nodes to show up in gossip, currently 1 present...
INFO  2023-08-16 14:35:13,200 [shard 0] compaction - [Compact system.local 1c185990-3c42-11ee-b4b5-76bbdda9abe9] Compacted 2 sstables to [/jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-8_abnlgw/test/node3/data/system/local-7ad54392bcdd35a684174e047860b377/mc-6-big-Data.db:level=0]. 13kB to 7kB (~55% of original) in 5ms = 2MB/s. ~256 total partitions merged to 1.
INFO  2023-08-16 14:35:16,130 [shard 0] storage_service - Set host_id=11ed336d-90ac-4543-bc62-64b4eb52e457 to be owned by node=127.0.80.2
INFO  2023-08-16 14:35:16,131 [shard 0] storage_service - Set host_id=6151ff93-067e-4fca-99ed-54abf9f7d028 to be owned by node=127.0.80.1
INFO  2023-08-16 14:35:18,202 [shard 0] gossip - (rate limiting dropped 454 similar messages) Waiting for 2 live nodes to show up in gossip, currently 1 present...
INFO  2023-08-16 14:35:23,208 [shard 0] gossip - (rate limiting dropped 453 similar messages) Waiting for 2 live nodes to show up in gossip, currently 1 present...
INFO  2023-08-16 14:35:28,218 [shard 0] gossip - (rate limiting dropped 453 similar messages) Waiting for 2 live nodes to show up in gossip, currently 1 present...
INFO  2023-08-16 14:35:33,221 [shard 0] gossip - (rate limiting dropped 452 similar messages) Waiting for 2 live nodes to show up in gossip, currently 1 present...
INFO  2023-08-16 14:35:38,230 [shard 0] gossip - (rate limiting dropped 453 similar messages) Waiting for 2 live nodes to show up in gossip, currently 1 present...
ERROR 2023-08-16 14:35:43,197 [shard 0] gossip - Timed out waiting for 2 live nodes to show up in gossip
```

While:
https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/6408/artifact/dtest/logs-full.release.005/1692196572172_counter_tests.py%3A%3ATestCountersOnMultipleNodes%3A%3Atest_counter_consistency_node_repair/node1.log
```
INFO  2023-08-16 14:35:01,092 [shard 0] init - Scylla version 5.4.0~dev-0.20230816.e9d94894f1d7 initialization completed.
INFO  2023-08-16 14:35:01,092 [shard 0] standard_role_manager - Created default superuser role 'cassandra'.
INFO  2023-08-16 14:35:04,728 [shard 0] compaction - [Compact system.scylla_local 170cb180-3c42-11ee-b026-adc1e6f218b9] Compacting [/jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-8_abnlgw/test/node1/data/system/scylla_local-2972ec7ffb2038ddaac1d876f2e3fcbd/me-3g8k_14ig_4aqzk2ofold0r71m0p-big-Data.db:level=0:origin=memtable,/jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-8_abnlgw/test/node1/data/system/scylla_local-2972ec7ffb2038ddaac1d876f2e3fcbd/me-3g8k_14ic_5x9gx2ofold0r71m0p-big-Data.db:level=0:origin=compaction]
INFO  2023-08-16 14:35:04,728 [shard 0] gossip - InetAddress 127.0.80.2 is now UP, status = UNKNOWN
INFO  2023-08-16 14:35:04,735 [shard 0] compaction - [Compact system.scylla_local 170cb180-3c42-11ee-b026-adc1e6f218b9] Compacted 2 sstables to [/jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-8_abnlgw/test/node1/data/system/scylla_local-2972ec7ffb2038ddaac1d876f2e3fcbd/me-3g8k_14ig_4c1a92ofold0r71m0p-big-Data.db:level=0]. 11kB to 6kB (~51% of original) in 5ms = 2MB/s. ~256 total partitions merged to 1.
INFO  2023-08-16 14:35:05,054 [shard 0] storage_service - bootstrap[c43a52fa-10de-4864-96ec-bd2817934b48]: Added node=127.0.80.2 as bootstrap, coordinator=127.0.80.2
INFO  2023-08-16 14:35:05,728 [shard 0] compaction - [Compact system.scylla_local 17a54800-3c42-11ee-b026-adc1e6f218b9] Compacting [/jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-8_abnlgw/test/node1/data/system/scylla_local-2972ec7ffb2038ddaac1d876f2e3fcbd/me-3g8k_14ih_4aqzk2ofold0r71m0p-big-Data.db:level=0:origin=memtable,/jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-8_abnlgw/test/node1/data/system/scylla_local-2972ec7ffb2038ddaac1d876f2e3fcbd/me-3g8k_14ig_4c1a92ofold0r71m0p-big-Data.db:level=0:origin=compaction]
INFO  2023-08-16 14:35:05,737 [shard 0] cdc - Starting to use generation (2023/08/16 14:35:30, 9eda088b-d6ab-4b1d-b8cb-e21048ad71f8)
INFO  2023-08-16 14:35:05,742 [shard 0] compaction - [Compact system.scylla_local 17a54800-3c42-11ee-b026-adc1e6f218b9] Compacted 2 sstables to [/jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-8_abnlgw/test/node1/data/system/scylla_local-2972ec7ffb2038ddaac1d876f2e3fcbd/me-3g8k_14ih_4c1a92ofold0r71m0p-big-Data.db:level=0]. 11kB to 6kB (~51% of original) in 9ms = 1MB/s. ~256 total partitions merged to 1.
INFO  2023-08-16 14:35:05,742 [shard 0] compaction - [Compact system.cdc_local 17a76ae0-3c42-11ee-b026-adc1e6f218b9] Compacting [/jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-8_abnlgw/test/node1/data/system/cdc_local-0bcaffd40c833eadad13dc1d5015b77c/me-3g8k_14ih_4cgps2ofold0r71m0p-big-Data.db:level=0:origin=memtable,/jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-8_abnlgw/test/node1/data/system/cdc_local-0bcaffd40c833eadad13dc1d5015b77c/me-3g8k_14id_007ps2ofold0r71m0p-big-Data.db:level=0:origin=memtable]
INFO  2023-08-16 14:35:05,743 [shard 0] cdc - CDC description table successfully updated with generation (2023/08/16 14:35:30, 9eda088b-d6ab-4b1d-b8cb-e21048ad71f8).
INFO  2023-08-16 14:35:05,753 [shard 0] compaction - [Compact system.cdc_local 17a76ae0-3c42-11ee-b026-adc1e6f218b9] Compacted 2 sstables to [/jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-8_abnlgw/test/node1/data/system/cdc_local-0bcaffd40c833eadad13dc1d5015b77c/me-3g8k_14ih_4f1b52ofold0r71m0p-big-Data.db:level=0]. 68kB to 63kB (~92% of original) in 9ms = 7MB/s. ~256 total partitions merged to 2.
INFO  2023-08-16 14:35:08,679 [shard 0] storage_service - bootstrap[c43a52fa-10de-4864-96ec-bd2817934b48]: Marked ops done from coordinator=127.0.80.2
INFO  2023-08-16 14:35:08,723 [shard 0] storage_service - Set host_id=11ed336d-90ac-4543-bc62-64b4eb52e457 to be owned by node=127.0.80.2
INFO  2023-08-16 14:35:15,126 [shard 0] compaction - [Compact system.scylla_local 1d3f4d60-3c42-11ee-b026-adc1e6f218b9] Compacting [/jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-8_abnlgw/test/node1/data/system/scylla_local-2972ec7ffb2038ddaac1d876f2e3fcbd/me-3g8k_14iq_15smo2ofold0r71m0p-big-Data.db:level=0:origin=memtable,/jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-8_abnlgw/test/node1/data/system/scylla_local-2972ec7ffb2038ddaac1d876f2e3fcbd/me-3g8k_14ih_4c1a92ofold0r71m0p-big-Data.db:level=0:origin=compaction]
INFO  2023-08-16 14:35:15,126 [shard 0] gossip - InetAddress 127.0.80.3 is now UP, status = UNKNOWN
INFO  2023-08-16 14:35:15,134 [shard 0] compaction - [Compact system.scylla_local 1d3f4d60-3c42-11ee-b026-adc1e6f218b9] Compacted 2 sstables to [/jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-8_abnlgw/test/node1/data/system/scylla_local-2972ec7ffb2038ddaac1d876f2e3fcbd/me-3g8k_14ir_0r0812ofold0r71m0p-big-Data.db:level=0]. 11kB to 6kB (~51% of original) in 6ms = 1MB/s. ~256 total partitions merged to 1.
INFO  2023-08-16 14:35:47,158 [shard 0] gossip - FatClient 127.0.80.3 has been silent for 30000ms, removing from gossip
```

https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/6408/artifact/dtest/logs-full.release.005/1692196572172_counter_tests.py%3A%3ATestCountersOnMultipleNodes%3A%3Atest_counter_consistency_node_repair/node2.log
```
INFO  2023-08-16 14:35:08,712 [shard 0] init - Scylla version 5.4.0~dev-0.20230816.e9d94894f1d7 initialization completed.
INFO  2023-08-16 14:35:08,718 [shard 0] compaction - [Compact system.peers 196b8910-3c42-11ee-be21-2383ed75a68d] Compacted 2 sstables to [/jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-8_abnlgw/test/node2/data/system/peers-37f71aca7dc2383ba70672528af04d4f/me-3g8k_14ik_473td2w36clqcey3yl-big-Data.db:level=0]. 18kB to 11kB (~65% of original) in 11ms = 1MB/s. ~256 total partitions merged to 1.
INFO  2023-08-16 14:35:15,126 [shard 0] compaction - [Compact system.scylla_local 1d3f4d60-3c42-11ee-be21-2383ed75a68d] Compacting [/jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-8_abnlgw/test/node2/data/system/scylla_local-2972ec7ffb2038ddaac1d876f2e3fcbd/me-3g8k_14iq_15smo2w36clqcey3yl-big-Data.db:level=0:origin=memtable,/jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-8_abnlgw/test/node2/data/system/scylla_local-2972ec7ffb2038ddaac1d876f2e3fcbd/me-3g8k_14ih_0f7u92w36clqcey3yl-big-Data.db:level=0:origin=compaction]
INFO  2023-08-16 14:35:15,126 [shard 0] gossip - InetAddress 127.0.80.3 is now UP, status = UNKNOWN
INFO  2023-08-16 14:35:15,133 [shard 0] compaction - [Compact system.scylla_local 1d3f4d60-3c42-11ee-be21-2383ed75a68d] Compacted 2 sstables to [/jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-8_abnlgw/test/node2/data/system/scylla_local-2972ec7ffb2038ddaac1d876f2e3fcbd/me-3g8k_14ir_0r0812w36clqcey3yl-big-Data.db:level=0]. 11kB to 6kB (~51% of original) in 6ms = 1MB/s. ~256 total partitions merged to 1.
INFO  2023-08-16 14:35:47,158 [shard 0] gossip - FatClient 127.0.80.3 has been silent for 30000ms, removing from gossip
```

**Comment by kbr-scylla** [comment]
Is that `test_counter_consistency_node_repair` failure related?

**Comment by bhalevy** [comment]
> Is that `test_counter_consistency_node_repair` failure related?

It's another example of a node timing out wait_for_live_nodes_to_show_up
while the other nodes see it is up but not vice verse.

**Comment by bhalevy** [comment]
Let's wait for #14801 to land first and see if we still hit `Timed out waiting for 2 live nodes to show up in gossip`.

**Comment by bhalevy** [comment]
Closing in favor of https://github.com/scylladb/scylladb/pull/15186
